subsys/net/pkt_filter/net_filter.c

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(net_filter, CONFIG_NET_FILTER_LOG_LEVEL);
+
+#include <zephyr/net/net_core.h>
+#include <zephyr/spinlock.h>
+#include <zephyr/net/net_ip.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/net/net_filter.h>
+
+struct nf_hook_entries {
+	sys_slist_t hook_root;
+	struct k_spinlock lock;
+};
+
+struct nf_hook_entry {
+	sys_snode_t node;
+	nf_hook_fn_t *hook_fn;
+	int priority;
+	struct net_if *iface;
+};
+
+#ifdef CONFIG_NET_IPV4
+static struct nf_hook_entries hooks_ipv4[NF_IP_NUMHOOKS];
+#endif
+
+#ifdef CONFIG_NET_IPV6
+static struct nf_hook_entries hooks_ipv6[NF_IP_NUMHOOKS];
+#endif
+
+static struct nf_hook_entries *nf_hook_entry_head(uint8_t pf, unsigned int hooknum)
+{
+	switch (pf) {
+	case PF_INET:
+	#ifdef CONFIG_NET_IPV4
+		if (ARRAY_SIZE(hooks_ipv4) <= hooknum) {
+			return NULL;
+		}
+		return &hooks_ipv4[hooknum];
+	#endif
+		break;
+	case PF_INET6:
+	#ifdef CONFIG_NET_IPV6
+		if (ARRAY_SIZE(hooks_ipv6) <= hooknum) {
+			return NULL;
+		}
+		return &hooks_ipv6[hooknum];
+	#endif
+		break;
+	default:
+		return NULL;
+	}
+
+	return NULL;
+}
+
+static void nf_lock_and_remove_hook(struct nf_hook_entries *entry,
+				    const struct nf_hook_cfg *cfg)
+{
+	struct nf_hook_entry *hook;
+	struct nf_hook_entry *tmp;
+	sys_snode_t *prev_node = NULL;
+	k_spinlock_key_t key = k_spin_lock(&entry->lock);
+
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&entry->hook_root, hook, tmp, node) {
+		if ((cfg->hook_fn == hook->hook_fn) && (cfg->priority == hook->priority)) {
+			sys_slist_remove(&entry->hook_root, prev_node, &hook->node);
+			k_free(hook);
+			NET_DBG("Unregistered hook %p [priority: %d]", hook->hook_fn, hook->priority);
+		} else {
+			prev_node = &hook->node;
+		}
+	}
+	k_spin_unlock(&entry->lock, key);
+}
+
+static void nf_lock_and_register_hook(struct nf_hook_entries *entry,
+				      struct nf_hook_entry *hook)
+{
+	k_spinlock_key_t key = k_spin_lock(&entry->lock);
+
+	if (sys_slist_is_empty(&entry->hook_root)) {
```

Comment:
Is this extra case really needed? I believe we could `sys_slist_prepend()` even on an empty list.

Reply:
Right, this additional case is not needed.